Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+
+            let bitpatternAsPointer = UnsafeMutableRawPointer.init(bitPattern: bitPattern)
+
+            _debugPrintCQE("\(i) = fd[\(fd)] eventType[\(String(describing:eventType))] sequenceNumber[\(sequenceNumber)] res [\(c.res)] flags [\(c.flags)]  bitpattern[\(String(describing:bitpatternAsPointer))]")
+        }
+        #endif
+    }
+
+    init() {
+        cqes = UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>.allocate(capacity: Int(cqeMaxCount))
+        cqes.initialize(repeating:&emptyCqe, count:Int(cqeMaxCount))
+    }
+
+    deinit {
+        cqes.deallocate()
+    }
+
+    internal func io_uring_queue_init() throws -> () {
+        // IORING_SETUP_SQPOLL will be fundamentally useful in 5.13, as we no longer need elevated privileges as previously
+        // and CNIOLinux.io_uring_queue_init will setup a shared uring instance which allows us to use a single
+        // kernel sqpoll thread that is shared amongst all loops. NB, each process will have it's own polling thread
+        // if SQPOLL is enabled, so for hosts running multiple processes using it, some care should be taken to not overload.
+        #if SWIFTNIO_LIBURING_SQPOLL_ENABLED
+        let _sqpollEnabled : UInt32 = 1
+        #else
+        let _sqpollEnabled : UInt32 = 0
+        #endif
+        if (CNIOLinux_io_uring_queue_init(ringEntries, &ring, _sqpollEnabled ) != 0)
+         {
+             throw UringError.uringSetupFailure
+         }
+
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+
+    internal func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux.io_uring_queue_exit(&ring)
+    }
+
+    static internal func io_uring_use_multishot_poll() -> Bool {
+        #if SWIFTNIO_LIBURING_MULTISHOT
+        return true
+        #else
+        return false
+        #endif
+    }
+
+    // Adopting some retry code from queue.c from liburing with slight
+    // modifications - we never want to have to handle retries of
+    // SQE allocation in all places it could possibly occur.
+    // If the SQ ring is full, we may need to submit IO first
+    internal func get_sqe() -> UnsafeMutablePointer<io_uring_sqe>?
+    {
+        // io_uring_submit can fail here due to backpressure from kernel for not reaping CQE:s.
+        //
+        // I think we should consider handling that as a fatalError, as fundamentally the ring size is too small
+        // compared to the amount of events the user tries to push through in a single eventloop tick.
+        //
+        // This is mostly a problem for synthetic tests that e.g. do a huge amount of registration modifications.
+        //
+        // This is a slight design issue with SwiftNIO in general that should be discussed.
+        //
+        while true {
+            if let sqe = CNIOLinux.io_uring_get_sqe(&ring) {
+                return sqe
+            }
+            self.io_uring_flush()
```

Comment:
is there a guarantee that this works and eventually gets space? Because we're not consuming CQE here, right?

Reply:
No, there is no guarantee - se comment above, I could see us doing a few retries and then fatalError() if failing. This is tied to the overall discussion above about the dynamics of generating outbound events / reaping events - if having a "bad producer" generating infinite events, we will eventually get back pressure from the kernel and not being able to send more. It's fundamentally an issue for the synthetic tests that do a lot of registration changes in a loop - should not be an issue for well behaved normal channel handlers I think. I think one possible solution would be to just keep track of all pending registration changes that needs to be sent to the kernel and merge it, and just do the actual re-registration to the kernel when we enter whenReady0 the next time. I've done that in the io_uring selector except for the merging (I postpone the flushing of registration changes until the next whenReady0 call, which helps alleviate this for real-world scenarios, but an aggressive enough client will eventually run out of SQE:s and block there.)